core/trino-main/src/main/java/io/trino/util/ReentrantResourceLock.java

```diff
@@ -0,0 +1,46 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.util;
+
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * ReentrantResourceLock has the same locking/unlocking semantics as ReentrantLock.
+ *
+ * The main difference is that this class is AutoCloseable which makes usage easier and
+ * consistent.
+ *
+ * <pre>{@code
+ *      ReentrantResourceLock lock = new ReentrantResourceLock();
+ *      try (ReentrantResourceLock ignored = lock.acquire()) {
+ *          // critical section
+ *      }
+ * }</pre>
+ */
+public final class ReentrantResourceLock
```

Comment:
Relevant discussions: https://github.com/google/guava/issues/1608

BTW, instead of a dedicated class (which imposes a choice of underlying lock implementation), it may be preferable to have a utility that wraps a lock:

```java
void testX()
{
    Lock lock = new ReentrantLock();
    try (UncheckedCloseable ignored = closeable(lock)) {
        // ...
    }
}

private interface UncheckedCloseable
    extends AutoCloseable
{
    void close();
}

private static UncheckedCloseable closeable(Lock lock)
{
    lock.lock();
    return lock::unlock;
}
```

Reply:
Also, it's weird, and possibly confusing, that `acquire()` returns a ReentrantResourceLock. A user might wonder if that's a new lock: should unlock be called on the original lock instance? On the one returned by the acquire method? What happens in either case?